File: linked list/160_Intersection of Two Linked Lists.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        dict = {}
        current = headA
        while current:
            dict[current] = current.val
            current = current.next
        current = headB
        while current:
            if current in dict:
                return current
            current = current.next
        return None

File: linked list/160_Intersection of Two Linked Lists.java
import java.util.HashMap;

class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
        next = null;
    }
}

class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        HashMap<ListNode, Integer> hashMap = new HashMap<>();
        ListNode current = headA;
        while (current != null) {
            hashMap.put(current, current.val);
            current = current.next;
        }
        current = headB;
        while (current != null) {
            if (hashMap.get(current) != null) {
                return current;
            }
            current = current.next;
        }

        return null;
    }
}
